test/proposal.test.ts

```diff
@@ -250,4 +254,122 @@ describe("proposal", function () {
       });
     });
   });
+
+  describe("if the following contract is deployed in advance, it still deploys", function () {
+    // We set up `earlierDeployment` so that for every deterministically
+    // computed address in the proposal we get a way to deterministically deploy
+    // that address as well as the expected deployment address.
+    // The function will be called in advance to predeploy the contract.
+    type DeployTransaction = (signer: Signer) => Promise<void>;
+    let earlierDeployments: Record<
+      DeterministicallyComputedAddress,
+      [string, DeployTransaction]
+    >;
+
+    // Change the salt to avoid conflicts with other tests (TODO: global fix!).
+    const modifiedSettings = {
+      ...settings,
+      cowDao: { ...settings.cowDao },
+      teamController: { ...settings.teamController },
+    };
+    modifiedSettings.cowDao.nonce = utils.hexZeroPad("0x31337", 32);
+    modifiedSettings.teamController.nonce = utils.hexZeroPad("0x31337", 32);
+
+    before(async function () {
+      async function deterministicDeploySafe({
+        owners,
+        threshold,
+        nonce,
+      }: SafeCreationSettings): Promise<[string, DeployTransaction]> {
+        const { to, data, address } = await prepareDeterministicSafeWithOwners(
+          owners,
+          threshold,
+          gnosisSafeManager.getDeploymentAddresses(),
+          BigNumber.from(nonce ?? 0),
+          ethers,
+        );
+        return [
+          address,
+          async function (deployer: Signer) {
+            await deployer.sendTransaction({ to, data });
+          },
+        ];
+      }
+
+      async function deterministicDeployCowToken(
+        params: Omit<RealTokenDeployParams, "totalSupply">,
+      ): Promise<[string, DeployTransaction]> {
+        const {
+          address,
+          safeTransaction: { to, data },
+        } = await getDeterministicDeploymentTransaction(
+          ContractName.RealToken,
+          {
+            totalSupply: BigNumber.from(10).pow(3 * 3 + metadata.real.decimals),
+            ...params,
+          },
+          ethers,
+        );
+        return [
+          address,
+          async function (deployer: Signer) {
+            await deployer.sendTransaction({ to, data });
+          },
+        ];
+      }
+
+      const cowDao = await deterministicDeploySafe(modifiedSettings.cowDao);
+      earlierDeployments = {
+        cowDao,
+        teamController: await deterministicDeploySafe(
+          modifiedSettings.teamController,
+        ),
+        investorFundsTarget: await deterministicDeploySafe({
+          threshold: 1,
+          owners: [cowDao[0]],
+        }),
+        cowToken: await deterministicDeployCowToken({
+          cowDao: cowDao[0],
+          initialTokenHolder: cowDao[0],
+        }),
+      };
+    });
+
+    for (const deterministicAddress of deterministicallyComputedAddresses) {
+      it(deterministicAddress, async function () {
```

Comment:
I like how you are automatically test it for every different determinsitic address. Cool.